File: ui_selenium_bdd/agents/advisor-agent.py
# ============================================================
# Advisor Agent — Décision Release · Prédiction · Recommandations
# ============================================================
# Commandes :
#   python agents/advisor-agent.py release [N]  → Go/No-Go production (N votes)
#   python agents/advisor-agent.py predict      → prédiction prochains échecs
#   python agents/advisor-agent.py recommend    → recommandations amélioration
#   python agents/advisor-agent.py report       → rapport HTML
# ============================================================

import sys, os, json, glob, time
sys.stdout.reconfigure(encoding="utf-8", errors="replace")
sys.path.insert(0, os.path.dirname(__file__))

import llm
from prompt_store import PromptStore

_ps = PromptStore()

def _fmt(template: str, **kw) -> str:
    result = template
    for key, val in kw.items():
        result = result.replace("{" + key + "}", str(val))
    return result

FRAMEWORK  = os.path.abspath(os.path.join(os.path.dirname(__file__), ".."))
ALLURE_DIR = os.path.join(FRAMEWORK, "target", "allure-results")
DOCS_DIR   = os.path.join(FRAMEWORK, "docs")
FLAKY_FILE = os.path.join(DOCS_DIR, "flaky-report.json")
BASELINE   = os.path.join(DOCS_DIR, "baseline.json")

R = "\033[31m"; G = "\033[32m"; Y = "\033[33m"; C = "\033[36m"
W = "\033[1m";  E = "\033[0m"

VOTE_SCHEMA = {
    "type": "object",
    "properties": {
        "verdict":   {"type": "string", "enum": ["GO", "NO-GO"]},
        "risk":      {"type": "string", "enum": ["low", "medium", "high", "critical"]},
        "blockers":  {"type": "array", "items": {"type": "string"}},
        "warnings":  {"type": "array", "items": {"type": "string"}},
        "reasoning": {"type": "string"},
    }
}

PREDICT_SCHEMA = {
    "type": "object",
    "properties": {
        "gate_will_pass":      {"type": "boolean"},
        "confidence":          {"type": "number"},
        "likely_failures":     {"type": "array", "items": {"type": "string"}},
        "likely_flaky":        {"type": "array", "items": {"type": "string"}},
        "risk_level":          {"type": "string", "enum": ["low", "medium", "high", "critical"]},
        "reasoning":           {"type": "string"},
        "recommended_actions": {"type": "array", "items": {"type": "string"}},
    }
}

RISK_COLORS = {"low": G, "medium": Y, "high": R, "critical": R}


def collect_context() -> dict:
    results = []
    for f in glob.glob(os.path.join(ALLURE_DIR, "*.json")):
        try:
            with open(f, encoding="utf-8") as fh:
                data = json.load(fh)
            if "name" in data and "status" in data:
                results.append(data)
        except Exception:
            pass
    total   = len(results)
    passed  = sum(1 for r in results if r.get("status") == "passed")
    failed  = sum(1 for r in results if r.get("status") == "failed")
    broken  = sum(1 for r in results if r.get("status") == "broken")
    skipped = sum(1 for r in results if r.get("status") == "skipped")
    pass_rate = round(passed / total * 100, 1) if total else 0
    fail_rate = round((failed + broken) / total * 100, 1) if total else 0

    flaky_tests = {}
    if os.path.exists(FLAKY_FILE):
        try:
            with open(FLAKY_FILE, encoding="utf-8") as f:
                flaky_tests = json.load(f).get("flaky", {})
        except Exception:
            pass

    failures = [r for r in results if r.get("status") in ("failed", "broken")]
    return dict(total=total, passed=passed, failed=failed, broken=broken,
                skipped=skipped, pass_rate=pass_rate, fail_rate=fail_rate,
                failures=failures, flaky_tests=flaky_tests)


def format_context(ctx: dict) -> str:
    lines = [
        f"Pass rate  : {ctx['pass_rate']}% ({ctx['passed']}/{ctx['total']})",
        f"Fail rate  : {ctx['fail_rate']}% ({ctx['failed']+ctx['broken']} échecs)",
        f"Ignorés    : {ctx['skipped']}",
        f"Flaky      : {len(ctx['flaky_tests'])} tests instables",
    ]
    if ctx["failures"]:
        lines.append("\nÉchecs :")
        for r in ctx["failures"][:5]:
            msg = r.get("statusDetails", {}).get("message", "?")[:60]
            lines.append(f"  - {r['name'][:50]} : {msg}")
    return "\n".join(lines)


# ── Release ────────────────────────────────────────────────────────────────────

def cmd_release(n_votes: int = 3):
    print(f"\n{W}ADVISOR — Release Decision (Self-Consistency, {n_votes} votes){E}")
    print(f"{Y}  {n_votes} appels LLM indépendants, vote majoritaire{E}\n")
    ctx     = collect_context()
    ctx_str = format_context(ctx)

    _tpl = _ps.get("release_vote") or (
        "Tu es QA Lead senior. Évalue si ce build Selenium BDD peut être déployé en production.\n\n"
        "{context_str}\n\n"
        "Critères de blocage :\n"
        "- Pass rate < 90%\n"
        "- Taux d'échec > 5%\n"
        "- Test critique ou smoke en échec\n"
        "- Test flaky critique présent\n\n"
        "Renvoie un verdict GO ou NO-GO avec le risque, les blockers et le raisonnement."
    )
    messages = [{"role": "user", "content": _fmt(_tpl, context_str=ctx_str)}]
    result   = llm.chat_self_consistent(messages, VOTE_SCHEMA, verdict_key="verdict", n=n_votes)
    _ps.record_usage("release_vote")

    verdict   = result["majority_verdict"]
    agreement = result["agreement_rate"]
    gate_color = G if verdict == "GO" else R
    print(f"  {gate_color}{W}  VERDICT : {verdict}  {E}")
    print(f"  Agreement : {int(agreement * 100)}%  |  Votes : {result['votes']}")

    majority_resp = next((r for r in result["responses"] if str(r.get("verdict","")).upper() == verdict), {})
    risk = majority_resp.get("risk", "unknown")
    print(f"  Risque    : {RISK_COLORS.get(risk,Y)}{risk.upper()}{E}")

    for b in majority_resp.get("blockers", []):
        print(f"  {R}✗{E} {b}")
    for w in majority_resp.get("warnings", []):
        print(f"  {Y}!{E} {w}")
    if agreement < 1.0:
        print(f"\n  {Y}[WARN] Désaccord ({int(agreement*100)}%). Vérification humaine recommandée.{E}")
    print(f"\n  {gate_color}Déploiement {'AUTORISÉ' if verdict=='GO' else 'BLOQUÉ'}.{E}")
    return verdict == "GO", result


def cmd_predict():
    print(f"\n{W}ADVISOR — Prédiction (Structured Output){E}\n")
    ctx     = collect_context()
    ctx_str = format_context(ctx)

    _tpl = _ps.get("predict_gate") or (
        "Analyse les tendances de cette suite de tests Selenium BDD et prédis :\n\n"
        "{context_str}\n\n"
        "Sur la base de ces données, prédis pour le PROCHAIN run :\n"
        "1. Est-ce que le quality gate va passer ?\n"
        "2. Quels tests vont probablement échouer ?\n"
        "3. Quels tests vont devenir flaky ?\n"
        "4. Quel est le niveau de risque global ?\n"
        "5. Quelles actions recommandes-tu avant le prochain run ?\n"
    )
    messages = [{"role": "user", "content": _fmt(_tpl, context_str=ctx_str)}]
    try:
        result = llm.chat_structured(messages, PREDICT_SCHEMA)
        _ps.record_usage("predict_gate", confidence=result.get("confidence", 0))
    except Exception as e:
        print(f"{R}  LLM indisponible : {e}{E}")
        return {}

    gate_ok    = result.get("gate_will_pass", False)
    confidence = result.get("confidence", 0)
    risk       = result.get("risk_level", "unknown")
    print(f"  Gate prochain run : {G if gate_ok else R}{'OUI' if gate_ok else 'NON'}{E} "
          f"(confidence : {int(confidence*100)}%)")
    print(f"  Risque global     : {RISK_COLORS.get(risk,Y)}{risk.upper()}{E}")

    if result.get("likely_failures"):
        print(f"\n  {R}Tests susceptibles d'échouer :{E}")
        for t in result["likely_failures"][:5]:
            print(f"    - {t}")
    if result.get("recommended_actions"):
        print(f"\n  {C}Actions recommandées :{E}")
        for a in result["recommended_actions"][:5]:
            print(f"    → {a}")
    return result


def cmd_recommend():
    print(f"\n{W}ADVISOR — Recommandations qualité{E}\n")
    ctx     = collect_context()
    ctx_str = format_context(ctx)
    messages = [{"role": "user", "content": (
        f"Analyse cette suite de tests Selenium BDD :\n\n{ctx_str}\n\n"
        f"Application testée : automationexercise.com (e-commerce, 26 cas de test)\n"
        f"Stack : Java 17, Selenium 4, Cucumber 7, TestNG\n\n"
        f"Donne 5 recommandations concrètes pour améliorer la qualité et la maintenabilité."
    )}]
    try:
        result = llm.chat(messages)
        for line in result.strip().split("\n"):
            print(f"  {line}")
    except Exception as e:
        print(f"{R}  LLM indisponible : {e}{E}")


def cmd_report():
    print(f"\n{W}ADVISOR — Rapport HTML{E}")
    ctx    = collect_context()
    stats  = {k: ctx[k] for k in ("total","passed","failed","broken","skipped","pass_rate","fail_rate")}
    gate_ok = stats["pass_rate"] >= 90 and stats["fail_rate"] <= 5
    gate_color = "#2ecc71" if gate_ok else "#e74c3c"

    html = f"""<!DOCTYPE html><html><head><meta charset="utf-8">
<title>Advisor Report — Selenium BDD</title>
<style>body{{background:#0d1117;color:#c9d1d9;font-family:'Segoe UI',sans-serif;padding:20px}}
.card{{background:#161b22;border:1px solid #30363d;border-radius:8px;padding:16px;margin:12px 0}}
h1{{color:#58a6ff}}h2{{color:#58a6ff;font-size:1em}}
.verdict{{font-size:2em;font-weight:bold;color:{gate_color};text-align:center;padding:20px}}
</style></head><body>
<h1>📊 Advisor Report — ui_selenium_bdd</h1>
<p style="color:#8b949e">Généré le {time.strftime('%Y-%m-%d %H:%M')}</p>
<div class="card"><div class="verdict">{'✅ GO — RELEASE AUTORISÉE' if gate_ok else '❌ NO-GO — RELEASE BLOQUÉE'}</div></div>
<div class="card"><h2>Métriques</h2>
<p>Total: {stats['total']} | Passés: {stats['passed']} | Échoués: {stats['failed']} | Brisés: {stats['broken']}</p>
<p>Pass rate: {stats['pass_rate']}% | Fail rate: {stats['fail_rate']}%</p>
</div></body></html>"""

    os.makedirs(DOCS_DIR, exist_ok=True)
    with open(os.path.join(DOCS_DIR, "advisor-report.html"), "w", encoding="utf-8") as f:
        f.write(html)
    print(f"  {G}Rapport : docs/advisor-report.html{E}")


if __name__ == "__main__":
    import argparse
    parser = argparse.ArgumentParser(description="Advisor Agent — Selenium BDD")
    parser.add_argument("command", choices=["release", "predict", "recommend", "report"])
    parser.add_argument("n_votes", nargs="?", type=int, default=3)
    args = parser.parse_args()

    if args.command == "release":   cmd_release(args.n_votes)
    elif args.command == "predict": cmd_predict()
    elif args.command == "recommend": cmd_recommend()
    elif args.command == "report":  cmd_report()

File: perf_k6_qacart/scripts/agents/advisor-agent.js
'use strict';
// ============================================================
// Advisor Agent — Décisions Release & Intelligence Prédictive (Performance)
// ============================================================
// Usage:
//   node scripts/agents/advisor-agent.js advise [N]      Go/No-Go release perf (N votes, défaut 3)
//   node scripts/agents/advisor-agent.js predict         Prédit les prochaines dégradations
//   node scripts/agents/advisor-agent.js gate            Quality Gate LLM
//   node scripts/agents/advisor-agent.js history <name>  Historique d'un scénario/métrique
//   node scripts/agents/advisor-agent.js memory seed     Injecte des données démo
//   node scripts/agents/advisor-agent.js memory recurring Dégradations récurrentes
//   node scripts/agents/advisor-agent.js report          Génère docs/advisor-report.html
//
// Guards: Self-consistency N votes (défaut 3, configurable)
// ============================================================
require('dotenv').config();
process.env.NODE_TLS_REJECT_UNAUTHORIZED = '0';

const fs     = require('fs');
const path   = require('path');
const llm    = require('./llm');
const tracer = require('./shared/tracer');
const memory      = require('./shared/memory-store');
const promptStore = require('./shared/prompt-store');

function fmt(template, vars) {
  let result = template;
  for (const [k, v] of Object.entries(vars)) {
    result = result.split('{' + k + '}').join(String(v ?? '?'));
  }
  return result;
}

const FRAMEWORK   = path.join(__dirname, '..', '..');
const REPORTS_DIR = path.join(FRAMEWORK, 'reports');
const DOCS_DIR    = path.join(FRAMEWORK, 'docs');
const SCENARIOS   = ['smoke', 'load', 'stress'];

fs.mkdirSync(DOCS_DIR, { recursive: true });

const DRY_RUN = process.argv.includes('--dry-run');
const N_VOTES = parseInt(process.argv.find(a => /^\d+$/.test(a)) || '3');

const G = '\x1b[32m', R = '\x1b[31m', Y = '\x1b[33m', C = '\x1b[36m', B = '\x1b[1m', E = '\x1b[0m';

// ── Contexte release ───────────────────────────────────────────────────────────
function collectContext() {
  const ctx = { scenarios: {}, total_breaches: 0, breaches: [] };
  for (const scenario of SCENARIOS) {
    const fp = path.join(REPORTS_DIR, `summary-${scenario}.json`);
    if (!fs.existsSync(fp)) continue;
    let summary;
    try { summary = JSON.parse(fs.readFileSync(fp, 'utf8')); } catch { continue; }

    const dur = summary.metrics?.http_req_duration || {};
    const checks = summary.metrics?.checks || { passes: 0, fails: 0 };
    const httpFailed = summary.metrics?.http_req_failed || { value: 0 };

    const breaches = [];
    for (const [metricName, metric] of Object.entries(summary.metrics || {})) {
      for (const [expr, wasBreached] of Object.entries(metric.thresholds || {})) {
        if (wasBreached) breaches.push({ metric: metricName, expr });
      }
    }

    ctx.scenarios[scenario] = {
      p95_ms: Math.round(dur['p(95)'] || 0),
      avg_ms: Math.round(dur.avg || 0),
      error_rate: Math.round((httpFailed.value || 0) * 1000) / 10,
      checks_pass_rate: (checks.passes + checks.fails) ? Math.round(checks.passes / (checks.passes + checks.fails) * 1000) / 10 : 0,
      breaches,
    };
    ctx.total_breaches += breaches.length;
    ctx.breaches.push(...breaches.map(b => ({ scenario, ...b })));
  }
  return ctx;
}

// ── ADVISE — Self-Consistency voting ──────────────────────────────────────────
const VOTE_SCHEMA = {
  verdict:   'GO | NO-GO',
  risk:      'low | medium | high | critical',
  blockers:  'array of string — points bloquants (vide si GO)',
  warnings:  'array of string — points à surveiller',
  reasoning: 'string — justification en 2-3 phrases',
};

async function cmdAdvise(nVotes = N_VOTES) {
  console.log(`\n${B}=== ADVISOR — RELEASE ADVISE PERF (${nVotes} votes) [${llm.MODEL}] ===${E}`);
  const ctx = collectContext();
  const agentSummary = memory.getAgentSummary('quality-agent', 3);

  const scenarioSummary = Object.entries(ctx.scenarios).map(([s, m]) => `${s}: p95=${m.p95_ms}ms err=${m.error_rate}% checks=${m.checks_pass_rate}%`).join(' | ');
  console.log(`  ${C}Contexte : ${scenarioSummary} | ${ctx.total_breaches} seuil(s) dépassé(s)${E}`);
  console.log(`  ${C}Self-consistency : ${nVotes} votes indépendants...${E}\n`);

  const breachDetail = ctx.breaches.slice(0,8).map(b => `- ${b.scenario}/${b.metric}: ${b.expr}`).join('\n') || 'Aucun';
  const _tpl = promptStore.get('release_vote') ||
    'Tu es un Release Manager expert en performance. Décide si ce build est prêt pour la production, du point de vue charge/latence.\n\n' +
    'Métriques par scénario:\n{scenario_summary}\n\n' +
    'Seuils dépassés ({failures_count}):\n{fail_detail}\n\n' +
    'Historique agent qualité:\n{agent_summary}\n\n' +
    'Contexte : QACart Todo tourne sur un dyno Heroku gratuit partagé (pas d\'infra dédiée) — ' +
    'une dégradation en stress (@stress) est attendue et non-bloquante ; un dépassement en smoke/load ' +
    'est plus préoccupant. Critères Go: smoke et load sans dépassement critique.';
  const prompt = fmt(_tpl, {
    scenario_summary: scenarioSummary, failures_count: ctx.total_breaches, fail_detail: breachDetail,
    agent_summary: agentSummary,
    pass_rate: ctx.scenarios.load?.checks_pass_rate ?? 'N/A', passed: 'N/A', total: 'N/A',
    broken: 0, skipped: 0, context_str: scenarioSummary,
    fail_rate: ctx.scenarios.load?.error_rate ?? 'N/A',
    flaky_count: 'N/A', critical_bugs: 'N/A', high_bugs: 'N/A', medium_bugs: 'N/A',
  });

  const messages = [{ role: 'user', content: prompt }];
  const span     = new tracer.Span('adviseSelfConsistent', prompt, llm.MODEL).begin();

  try {
    const { responses, winner, vote_counts, n_votes } = await llm.chatSelfConsistent(messages, VOTE_SCHEMA, nVotes);
    span.end(true);
    promptStore.recordUsage('release_vote');

    console.log(`  ${B}Votes (${n_votes}/${nVotes}) :${E}`);
    responses.forEach((r, i) => {
      const icon = r.verdict === 'GO' ? G+'GO'+E : R+'NO-GO'+E;
      console.log(`  Vote ${i+1}: ${icon}  risque=${r.risk}  ${r.reasoning?.slice(0,60)||''}`);
    });

    if (winner) {
      const verdictIcon = winner.verdict === 'GO' ? G + '✅ GO — RELEASE AUTORISÉE' + E : R + '❌ NO-GO — BLOCAGE RELEASE' + E;
      console.log(`\n  ${B}Décision finale :${E} ${verdictIcon}`);
      console.log(`  Risque : ${winner.risk?.toUpperCase()}`);
      if (winner.blockers?.length) { console.log(`  ${R}Bloquants :${E}`); winner.blockers.forEach(b => console.log(`    • ${b}`)); }
      if (winner.warnings?.length) { console.log(`  ${Y}Warnings :${E}`);  winner.warnings.forEach(w => console.log(`    • ${w}`)); }
      console.log(`\n  ${C}${winner.reasoning||''}${E}`);

      memory.recordEpisode('advisor-agent', [{ verdict: winner.verdict, risk: winner.risk, votes: vote_counts }], `Release: ${winner.verdict} (risque ${winner.risk})`, 'advise');
    }
  } catch (e) {
    span.error = e.message; span.end(false);
    console.error(`  ${R}✗ ${e.message}${E}`);
  }
}

// ── PREDICT — Prédiction des prochaines dégradations ──────────────────────────
const PREDICT_SCHEMA = {
  tc:              'string — identifiant scénario/métrique',
  failure_risk:    'float 0.0-1.0 — probabilité de dépassement futur',
  trend:           'stable | degrading | improving',
  predicted_cause: 'string — cause probable',
  recommendation:  'string — action préventive',
};

async function cmdPredict() {
  console.log(`\n${B}=== ADVISOR — PREDICT [${llm.MODEL}] ===${E}`);
  const recurring = memory.getRecurringFailures(2);
  const episodes  = memory.loadAllEpisodes(null, 20);

  if (!Object.keys(recurring).length && !episodes.length) {
    console.log(`  ${Y}⚠  Pas assez d'historique — lance d'abord quelques runs${E}`); return;
  }

  console.log(`  ${C}${Object.keys(recurring).length} métriques récurrentes analysées...${E}\n`);

  for (const [tc, stats] of Object.entries(recurring).slice(0, 10)) {
    const context = memory.getContextFor(tc);
    const span    = new tracer.Span('predictStructured', tc, llm.MODEL).begin();

    try {
      const _tpl = promptStore.get('predict_gate') ||
        'Prédit le risque de dépassement futur pour cette métrique k6.\n\n' +
        'Métrique : {test_name}\nHistorique :\n{context}\n' +
        'Statistiques : {count} occurrences, catégorie dominante: {dominant_category}';
      const prompt = fmt(_tpl, {
        test_name: tc, context, count: stats.count,
        dominant_category: stats.dominant_category,
        pass_rate: 'N/A', passed: 'N/A', total: 'N/A',
        failures_count: stats.count, broken: 0, history_summary: context,
      });

      const pred = await llm.chatStructured([{ role: 'user', content: prompt }], PREDICT_SCHEMA);
      span.end(true);
      promptStore.recordUsage('predict_gate', pred.failure_risk);

      const riskPct = Math.round((pred.failure_risk || 0) * 100);
      const riskIcon = riskPct >= 70 ? R+'⚠ HIGH'+E : riskPct >= 40 ? Y+'~ MED'+E : G+'✓ LOW'+E;
      console.log(`  ${riskIcon}  ${tc.slice(0,45).padEnd(45)}  ${riskPct}%  ${pred.trend||'?'}  ${pred.recommendation?.slice(0,40)||''}`);
    } catch (e) {
      span.error = e.message; span.end(false);
      console.error(`  ${R}✗ ${tc}: ${e.message}${E}`);
    }
  }
}

// ── GATE — Quality Gate LLM ───────────────────────────────────────────────────
async function cmdGate() {
  console.log(`\n${B}=== ADVISOR — GATE [${llm.MODEL}] ===${E}`);
  const ctx  = collectContext();
  const span = new tracer.Span('gateStructured', JSON.stringify(ctx), llm.MODEL).begin();

  const GATE_SCHEMA = {
    verdict:     'PASS | FAIL',
    score:       'integer 0-100',
    blockers:    'array of string',
    improvement: 'string — suggestion principale',
  };

  try {
    const scenarioSummary = Object.entries(ctx.scenarios).map(([s, m]) => `${s}: p95=${m.p95_ms}ms err=${m.error_rate}%`).join(' | ');
    const prompt = `Évalue la qualité performance de ce build k6 et donne un score 0-100.

${scenarioSummary}
Seuils dépassés: ${ctx.total_breaches}

Critères: 0 dépassement en smoke/load → PASS, dépassement uniquement en stress (dyno gratuit partagé) → WARNING, dépassement en smoke/load → FAIL`;

    const gate = await llm.chatStructured([{ role: 'user', content: prompt }], GATE_SCHEMA);
    span.end(true);

    const icon = gate.verdict === 'PASS' ? G+'✅ PASS'+E : R+'❌ FAIL'+E;
    console.log(`\n  ${icon}  Score : ${gate.score}/100`);
    if (gate.blockers?.length) gate.blockers.forEach(b => console.log(`  ${R}• ${b}${E}`));
    if (gate.improvement) console.log(`  ${C}→ ${gate.improvement}${E}`);
  } catch (e) {
    span.error = e.message; span.end(false);
    console.error(`  ${R}✗ ${e.message}${E}`);
  }
}

// ── HISTORY ───────────────────────────────────────────────────────────────────
function cmdHistory(tcId) {
  if (!tcId) { console.log(`  Usage: advisor history <scenario/metric>`); return; }
  console.log(`\n${B}=== ADVISOR — HISTORY : ${tcId} ===${E}`);
  const hist = memory.getTcHistory(tcId, 10);
  if (!hist.length) { console.log(`  ${Y}⚠  Aucun historique pour ${tcId}${E}`); return; }
  console.log(`  ${C}${hist.length} entrées dans la mémoire épisodique${E}\n`);
  hist.forEach(h => {
    const icon = (h.category === 'backend_degradation' || h.verdict === 'NO-GO') ? R+'✗'+E : G+'✓'+E;
    console.log(`  ${icon} ${h.ts.slice(0,10)}  ${h.agent.padEnd(20)}  ${h.category||h.verdict||'?'}  conf:${h.confidence!=null?Math.round(h.confidence*100)+'%':'?'}`);
  });
  console.log(`\n  ${memory.getContextFor(tcId)}`);
}

// ── MEMORY ────────────────────────────────────────────────────────────────────
function cmdMemory(action) {
  if (action === 'recurring') {
    console.log(`\n${B}=== ADVISOR — MÉMOIRE RÉCURRENTE ===${E}`);
    const rec = memory.getRecurringFailures(2);
    if (!Object.keys(rec).length) { console.log(`  ${Y}⚠  Aucune dégradation récurrente${E}`); return; }
    Object.entries(rec).forEach(([tc, s]) => {
      console.log(`  ${R}⚠${E}  ${tc.slice(0,50).padEnd(50)}  ×${s.count}  ${s.dominant_category}  last: ${s.last_seen.slice(0,10)}`);
    });
  } else if (action === 'seed') {
    console.log(`\n${B}=== ADVISOR — SEED MÉMOIRE ===${E}`);
    const demos = [
      { tc: 'stress/http_req_duration', category: 'env_shared_demo', confidence: 0.82 },
      { tc: 'load/http_req_failed',     category: 'test_config_too_aggressive', confidence: 0.71 },
    ];
    for (const d of demos) {
      memory.recordEpisode('quality-agent', [d], `Demo: ${d.category}`, 'seed');
      console.log(`  ${G}✓${E} ${d.tc}`);
    }
  } else {
    const eps = memory.loadAllEpisodes(null, 10);
    console.log(`\n${B}=== ADVISOR — MÉMOIRE (${eps.length} épisodes) ===${E}`);
    eps.slice(-10).reverse().forEach(ep => {
      console.log(`  ${ep.ts.slice(0,10)}  ${ep.agent.padEnd(20)}  ${ep.summary?.slice(0,50)||''}`);
    });
  }
}

// ── REPORT ────────────────────────────────────────────────────────────────────
function cmdReport() {
  const ctx      = collectContext();
  const episodes = memory.loadAllEpisodes(null, 20);
  const recurring = memory.getRecurringFailures(2);

  const html = `<!DOCTYPE html>
<html lang="fr"><head><meta charset="UTF-8"><title>Advisor Report — k6 Performance</title>
<style>body{font-family:sans-serif;background:#0f172a;color:#e2e8f0;padding:24px}
h1{color:#7c3aed}h2{color:#94a3b8;font-size:1rem;margin-top:24px}
.card{background:#1e293b;border-radius:8px;padding:16px;margin:12px 0}
.go{color:#4ade80}.nogo{color:#f87171}.warn{color:#fbbf24}
table{width:100%;border-collapse:collapse}td,th{padding:8px;border-bottom:1px solid #1e293b}
th{color:#64748b;text-align:left}</style></head>
<body>
<h1>🧠 Advisor Report Performance — ${new Date().toLocaleDateString('fr-FR')}</h1>
<div class="card"><h2>Contexte Build</h2>
<table><tr><th>Scénario</th><th>p95</th><th>Err rate</th><th>Checks</th></tr>
${Object.entries(ctx.scenarios).map(([s,m])=>`<tr><td>${s}</td><td>${m.p95_ms}ms</td><td>${m.error_rate}%</td><td>${m.checks_pass_rate}%</td></tr>`).join('')}
</table></div>
<div class="card"><h2>Dégradations Récurrentes (≥2 occurrences)</h2>
<table><tr><th>Métrique</th><th>Occurrences</th><th>Catégorie</th><th>Dernier</th></tr>
${Object.entries(recurring).map(([tc,s])=>`<tr><td>${tc}</td><td class="nogo">${s.count}</td><td>${s.dominant_category}</td><td>${s.last_seen.slice(0,10)}</td></tr>`).join('')||'<tr><td colspan="4" style="color:#4ade80">Aucune</td></tr>'}
</table></div>
<div class="card"><h2>Historique Épisodique (20 derniers)</h2>
<table><tr><th>Date</th><th>Agent</th><th>Résumé</th></tr>
${episodes.slice(-20).reverse().map(ep=>`<tr><td>${ep.ts.slice(0,10)}</td><td>${ep.agent}</td><td>${ep.summary?.slice(0,60)||''}</td></tr>`).join('')}
</table></div>
</body></html>`;

  const outPath = path.join(DOCS_DIR, 'advisor-report.html');
  if (!DRY_RUN) { fs.writeFileSync(outPath, html, 'utf8'); console.log(`  ${G}✓ ${outPath}${E}`); }
}

// ── Main ──────────────────────────────────────────────────────────────────────
async function main() {
  const args = process.argv.slice(2).filter(a => !a.startsWith('--') && !/^\d+$/.test(a));
  const [cmd, sub] = args;
  await llm.assertRunning();

  switch (cmd) {
    case 'advise':  await cmdAdvise(N_VOTES); break;
    case 'predict': await cmdPredict();       break;
    case 'gate':    await cmdGate();          break;
    case 'history': cmdHistory(sub);          break;
    case 'memory':  cmdMemory(sub);           break;
    case 'report':  cmdReport();              break;
    default:
      console.log(`
${B}Advisor Agent${E} — Décisions Release & Intelligence Prédictive (Performance)

  advise [N]           Go/No-Go release perf par self-consistency (N votes, défaut 3)
  predict               Prédit les prochaines dégradations depuis la mémoire épisodique
  gate                  Quality Gate avec score LLM (0-100)
  history <scénario>    Historique complet d'un scénario/métrique dans la mémoire
  memory                Affiche les épisodes mémorisés
  memory recurring      Dégradations récurrentes (≥2 occurrences)
  memory seed            Injecte des données de démo
  report                 Génère docs/advisor-report.html

Options:
  --dry-run    Simulation sans écriture
  N            Nombre de votes pour advise (ex: advisor advise 5)
`);
  }
}

main().catch(e => { console.error(R + e.message + E); process.exit(1); });
